src/transport/iroh_net.rs

```diff
@@ -0,0 +1,750 @@
+                            connection.remote_address()
+                        )
+                    })
+                else {
+                    connection.close(0u32.into(), b"failed to extract iroh-net node id");
+                    continue;
+                };
+
+                if !allowed_node_ids.contains(&client_node_id) {
+                    connection.close(0u32.into(), b"forbidden node id");
+                    continue;
+                }
+            }
+
+            tracing::debug!(
+                "Connection established from {:?}",
+                connection.remote_address()
+            );
+
+            tracing::debug!("Spawning connection handler...");
+            tokio::spawn(Self::connection_handler(connection, sender.clone()));
+        }
+    }
+
+    /// Create a new server channel, given a quinn endpoint, with unrestricted access by node id
+    ///
+    /// The server channel will take care of listening on the endpoint and spawning
+    /// handlers for new connections.
+    pub fn new(endpoint: iroh_net::Endpoint) -> io::Result<Self> {
+        Self::new_with_access_control(endpoint, AccessControl::Unrestricted)
+    }
+
+    /// Create a new server endpoint, with specified access control
+    ///
+    /// The server channel will take care of listening on the endpoint and spawning
+    /// handlers for new connections.
+    pub fn new_with_access_control(
+        endpoint: iroh_net::Endpoint,
+        access_control: AccessControl,
+    ) -> io::Result<Self> {
+        let allowed_node_ids = match access_control {
+            AccessControl::Unrestricted => BTreeSet::new(),
+            AccessControl::Allowed(list) if list.is_empty() => {
+                return Err(io::Error::other(
+                    "Empty list of allowed nodes, \
+                    endpoint would reject all connections",
+                ));
+            }
+            AccessControl::Allowed(list) => BTreeSet::from_iter(list),
+        };
+
+        let (ipv4_socket_addr, maybe_ipv6_socket_addr) = endpoint.bound_sockets();
+        let (sender, receiver) = flume::bounded(16);
+        let task = tokio::spawn(Self::endpoint_handler(
+            endpoint.clone(),
+            sender,
+            allowed_node_ids,
+        ));
+
+        Ok(Self {
+            inner: Arc::new(ListenerInner {
+                endpoint: Some(endpoint),
+                task: Some(task),
+                local_addr: once(LocalAddr::Socket(ipv4_socket_addr))
+                    .chain(maybe_ipv6_socket_addr.map(LocalAddr::Socket))
+                    .collect(),
+                receiver,
+            }),
+            _p: PhantomData,
+        })
+    }
+
+    /// Create a new server channel, given just a source of incoming connections
+    ///
+    /// This is useful if you want to manage the quinn endpoint yourself,
+    /// use multiple endpoints, or use an endpoint for multiple protocols.
+    pub fn handle_connections(
```

Comment:
Very nice, that is exactly how I would have done it myself.

Reply:
Well, you probably did, because this part I take from the quinn transport LMAO, tried to keep the same API, just adapting what was needed for using iroh-net.